qubit-sdk/src/main/java/com/qubit/android/sdk/internal/placement/interactor/PlacementQueryAttributesBuilder.kt

```diff
@@ -18,71 +29,87 @@ internal class PlacementQueryAttributesBuilder {
       cachedAttributes: Map<String, JsonObject>
   ) = JsonObject().apply {
     add(VISITOR_ATTRIBUTE_KEY, buildVisitorAttributesJson(deviceId))
-    add(USER_ATTRIBUTE_KEY, customAttributes, cachedAttributes,
-        { convertToUserAttribute(it) }, { buildEmptyUserAttribute() })
-    add(VIEW_ATTRIBUTE_KEY, customAttributes, cachedAttributes,
-        { convertToViewAttribute(it) }, { buildEmptyViewAttribute() })
+    add(USER_ATTRIBUTE_KEY, USER_ATTRIBUTE_SCHEMA, customAttributes, cachedAttributes)
+    add(VIEW_ATTRIBUTE_KEY, VIEW_ATTRIBUTE_SCHEMA, customAttributes, cachedAttributes)
+    addCustomAttributes(customAttributes)
+  }
 
-    customAttributes?.keySet()
-        ?.filter { !has(it) }
-        ?.forEach { add(it, customAttributes.get(it)) }
+  internal fun buildVisitorAttributesJson(deviceId: String) = JsonObject().apply {
+    addProperty("id", deviceId)
+//    addProperty("userAgentString", "")  // TODO set value
   }
 
   private fun JsonObject.add(
       key: String,
+      schema: List<AttributeProperty>,
       customAttributes: JsonObject?,
-      cachedAttributes: Map<String, JsonObject>,
-      converter: (JsonObject) -> JsonObject,
-      emptyBuilder: () -> JsonObject
+      cachedAttributes: Map<String, JsonObject>
   ) {
-    val value = (customAttributes?.get(key) as? JsonObject)
-        // custom attributes passed by SDK user - should contain all the fields from schema and nothing more
-        ?.let { converter(it) }
-    // cached attributes - skipped if set by SDK user
-        ?: cachedAttributes[key]
-        // if expected attribute is missing, then we should add empty ones
-        ?: emptyBuilder()
-    add(key, value)
+    val valueJson = JsonObject().apply {
+      addCustomProperties(schema, customAttributes?.get(key) as? JsonObject)
+      addCachedProperties(schema, cachedAttributes[key])
```

Comment:
The same here. Name could be more descriptive. It's not clear, whether this method overwrites properties or not.